pkg/kubelet/status/manager.go

```diff
@@ -150,6 +154,34 @@ func (m *manager) SetPodStatus(pod *api.Pod, status api.PodStatus) {
 	m.updateStatusInternal(pod, status)
 }
 
+func (m *manager) ReconcilePodStatus(pod *api.Pod) {
+	m.podStatusesLock.Lock()
+	defer m.podStatusesLock.Unlock()
+
+	// The pod could be a mirror pod, so we should translate first.
+	uid := m.podManager.TranslatePodUID(pod.UID)
+	status, found := m.podStatuses[uid]
+	if !found {
+		// NOTE: (random-liu) What if no status? Is there any chance that this would happen?
+		glog.Warningf("Container to be reconciled has been removed in status cache: %q", format.Pod(pod))
+		return
+	}
+	if isStatusEqual(&status.status, &pod.Status) {
+		// If the status from the source is the same with the cached status,
+		// reconcile is not needed. Just return.
+		return
+	}
+	if m.needsUpdate(pod.UID, status) {
+		// The pod is going to be updated, the conflict will be reconciled after updating
+		return
+	}
+	// We should trigger an reconcile here
+	// After deleting the corresponding apiStatusVersions, the pod status will be updated in next syncBatch
+	// NOTE: (random-liu) Is apiStatusVersions-1 or status version +1 better?
+	delete(m.apiStatusVersions, pod.UID)
```

Comment:
I am still debating between a channel versus a lock. I like the cleanness of using a channel, but it also means that we may send more updates when there are pending status updates (for the same pod) in the main update channel.

The sync goroutine could:
 1. Clear the entire update channel, reset the apiserver version, and send the update again.
 2  Runs syncBatch which nullifies all updates queued in the update channel, reset the apiserver version, and send the update again.
 3. Resets apisever version first, and does either syncBatch or the process the main update channel 
 ...

Both (1) and (2) incur more updates.

If we use a lock and reset the apiserverVersion right away, any update processed after calling the Reconcile function would correctly set the apiserverVersion again without causing an extra update.

It might also be the case that this happens rarely and my concerns are irrelevant.